src/Xamarin.Android.Build.Tasks/Utilities/AssemblyStoreGenerator.cs

```diff
@@ -1,133 +1,343 @@
+		ulong namesSize = 0;
 
-			this.archiveAssembliesPrefix = archiveAssembliesPrefix;
-			this.log = log;
+		foreach (AssemblyStoreAssemblyInfo info in infos) {
+			namesSize += (ulong)info.AssemblyNameBytes.Length;
+			namesSize += sizeof (uint);
 		}
 
-		public void Add (string apkName, AssemblyStoreAssemblyInfo storeAssembly)
-		{
-			if (String.IsNullOrEmpty (apkName)) {
-				throw new ArgumentException ("must not be null or empty", nameof (apkName));
-			}
+		ulong assemblyDataStart = (infoCount * IndexEntrySize () * 2) + (AssemblyStoreEntryDescriptor.NativeSize * infoCount) + AssemblyStoreHeader.NativeSize + namesSize;
+		// We'll start writing to the stream after we seek to the position just after the header, index, descriptors and name data.
+		ulong curPos = assemblyDataStart;
 
-			Store store;
-			if (!stores.ContainsKey (apkName)) {
-				store = new Store {
-					Common = new CommonAssemblyStore (apkName, archiveAssembliesPrefix, log, GetNextStoreID (apkName), globalIndexCounter),
-					Arch = new ArchAssemblyStore (apkName, archiveAssembliesPrefix, log, GetNextStoreID (apkName), globalIndexCounter)
-				};
+		using var fs = File.Open (storePath, FileMode.Create, FileAccess.Write, FileShare.Read);
+		fs.Seek ((long)curPos, SeekOrigin.Begin);
 
-				stores.Add (apkName, store);
-				SetIndexStore (store.Common);
-				SetIndexStore (store.Arch);
-			}
+		foreach (AssemblyStoreAssemblyInfo info in infos) {
+			(AssemblyStoreEntryDescriptor desc, curPos) = MakeDescriptor (info, curPos);
+			desc.mapping_index = (uint)descriptors.Count;
+			descriptors.Add (desc);
 
-			store = stores[apkName];
-			if (String.IsNullOrEmpty (storeAssembly.Abi)) {
-				store.Common.Add (storeAssembly);
-			} else {
-				store.Arch.Add (storeAssembly);
+			if ((uint)fs.Position != desc.data_offset) {
+				throw new InvalidOperationException ($"Internal error: corrupted store '{storePath}' stream");
 			}
 
-			void SetIndexStore (AssemblyStore b)
-			{
-				if (!b.IsIndexStore) {
-					return;
-				}
+			ulong name_with_ext_hash = MonoAndroidHelper.GetXxHash (info.AssemblyNameBytes, is64Bit);
+			ulong name_no_ext_hash = MonoAndroidHelper.GetXxHash (info.AssemblyNameNoExtBytes, is64Bit);
+			index.Add (new AssemblyStoreIndexEntry (info.AssemblyName, name_with_ext_hash, desc.mapping_index));
+			index.Add (new AssemblyStoreIndexEntry (info.AssemblyNameNoExt, name_no_ext_hash, desc.mapping_index));
+
+			CopyData (info.SourceFile, fs, storePath);
+			CopyData (info.SymbolsFile, fs, storePath);
+			CopyData (info.ConfigFile, fs, storePath);
+		}
+		fs.Flush ();
+		fs.Seek (0, SeekOrigin.Begin);
 
-				if (indexStore != null) {
-					throw new InvalidOperationException ("Index store already set!");
-				}
+		uint storeVersion = is64Bit ? ASSEMBLY_STORE_FORMAT_VERSION_64BIT : ASSEMBLY_STORE_FORMAT_VERSION_32BIT;
+		var header = new AssemblyStoreHeader (storeVersion | abiFlag, infoCount, (uint)index.Count, (uint)(index.Count * IndexEntrySize ()));
```

Comment:
I don't understand `(uint)(index.Count * IndexEntrySize()`, used for the `index_entry_count` value.  I'd *expect* this to be `index.Count*2`, but `IndexEntrySize()` is either `8` (32-bit platforms) or `32` (64-bit platforms)`, which suggests I don't actually understand what `index_entry_count` is supposed to *be*.

This impacts the commit message File Formats section. :-)

Reply:
`IndexEntrySize()` returns the hash width

Reply:
The issue is I misread; `(uint) index.Count` is `index_entry_count`, while `(uint)(index.Count * IndexEntrySize ()` is `index_size`.  Doh!